File: python/packages/kagent-core/src/kagent/core/_structured_object.py
from __future__ import annotations

import json
from collections.abc import Mapping
from typing import Any

from google.protobuf import json_format, struct_pb2
from kagent.api.v1alpha1 import common_pb2


def encode_structured_object(
    value: Mapping[str, Any],
    *,
    api_version: str,
    kind: str,
    max_bytes: int,
) -> common_pb2.StructuredObject:
    if not kind:
        raise ValueError("structured object kind is empty")
    payload = dict(value)
    _check_size(payload, max_bytes)
    protobuf_value = struct_pb2.Struct()
    json_format.ParseDict(payload, protobuf_value)
    return common_pb2.StructuredObject(api_version=api_version, kind=kind, value=protobuf_value)


def decode_structured_object(
    value: common_pb2.StructuredObject,
    *,
    expected_kind: str,
    max_bytes: int,
) -> dict[str, Any]:
    if value is None or not value.HasField("value"):
        raise ValueError("structured object value is missing")
    if expected_kind and value.kind != expected_kind:
        raise ValueError(f"structured object kind {value.kind!r} does not match {expected_kind!r}")
    payload = json_format.MessageToDict(value.value)
    if not isinstance(payload, dict):
        raise ValueError("structured object root is not an object")
    _check_size(payload, max_bytes)
    return payload


def _check_size(value: Mapping[str, Any], max_bytes: int) -> None:
    encoded = json.dumps(value, ensure_ascii=False, allow_nan=False, separators=(",", ":")).encode()
    if max_bytes > 0 and len(encoded) > max_bytes:
        raise ValueError(
            f"structured object exceeds configured size limit: got {len(encoded)} bytes, limit {max_bytes}"
        )

File: go/api/structuredobject/structuredobject.go
package structuredobject

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"

	apiv1alpha1 "github.com/kagent-dev/kagent/go/api/gen/kagent/api/v1alpha1"
	"google.golang.org/protobuf/types/known/structpb"
)

var (
	ErrEmptyKind     = errors.New("structured object kind is empty")
	ErrKindMismatch  = errors.New("structured object kind does not match expected kind")
	ErrNilValue      = errors.New("structured object value is nil")
	ErrNonObjectRoot = errors.New("structured object JSON root is not an object")
	ErrValueTooLarge = errors.New("structured object exceeds configured size limit")
)

func FromGo(value any, apiVersion, kind string, maxBytes int) (*apiv1alpha1.StructuredObject, error) {
	if kind == "" {
		return nil, ErrEmptyKind
	}

	jsonValue, err := json.Marshal(value)
	if err != nil {
		return nil, fmt.Errorf("marshal structured object: %w", err)
	}
	if maxBytes > 0 && len(jsonValue) > maxBytes {
		return nil, fmt.Errorf("%w: got %d bytes, limit %d", ErrValueTooLarge, len(jsonValue), maxBytes)
	}

	var decoded any
	decoder := json.NewDecoder(bytes.NewReader(jsonValue))
	decoder.UseNumber()
	if err := decoder.Decode(&decoded); err != nil {
		return nil, fmt.Errorf("decode structured object: %w", err)
	}
	object, ok := decoded.(map[string]any)
	if !ok || object == nil {
		return nil, ErrNonObjectRoot
	}

	valueStruct, err := structpb.NewStruct(object)
	if err != nil {
		return nil, fmt.Errorf("create protobuf struct: %w", err)
	}
	return &apiv1alpha1.StructuredObject{
		ApiVersion: apiVersion,
		Kind:       kind,
		Value:      valueStruct,
	}, nil
}

func ToGo(object *apiv1alpha1.StructuredObject, expectedKind string, destination any, maxBytes int) error {
	if object == nil || object.GetValue() == nil {
		return ErrNilValue
	}
	if expectedKind != "" && object.GetKind() != expectedKind {
		return fmt.Errorf("%w: got %q, want %q", ErrKindMismatch, object.GetKind(), expectedKind)
	}

	jsonValue, err := object.GetValue().MarshalJSON()
	if err != nil {
		return fmt.Errorf("marshal protobuf struct: %w", err)
	}
	if maxBytes > 0 && len(jsonValue) > maxBytes {
		return fmt.Errorf("%w: got %d bytes, limit %d", ErrValueTooLarge, len(jsonValue), maxBytes)
	}

	decoder := json.NewDecoder(bytes.NewReader(jsonValue))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(destination); err != nil {
		return fmt.Errorf("decode structured object into destination: %w", err)
	}
	if err := ensureEOF(decoder); err != nil {
		return err
	}
	return nil
}

func ensureEOF(decoder *json.Decoder) error {
	var trailing any
	if err := decoder.Decode(&trailing); !errors.Is(err, io.EOF) {
		if err == nil {
			return errors.New("decode structured object: trailing JSON value")
		}
		return fmt.Errorf("decode structured object trailing data: %w", err)
	}
	return nil
}
